### PR TITLE
2.4.0 Updates to DCP

### PR DESCRIPTION
**Pull request by LucasLLC**
## Description
This PR updates tutorials to include:

- Asynchronous checkpointing examples
- Updates save examples to include Stateful behavior
- Adds a note about format utilities

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

**Review by wz337 — COMMENTED** [review]
We should probably add an item in the recipes_index.rst as well. 
https://github.com/wz337/tutorials/blob/main/recipes_source/recipes_index.rst

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 1]
```diff
@@ -0,0 +1,259 @@
+Asynchronous Saving with Distributed Checkpoint (DCP)
+=====================================================
+
+Checkpointing is often a bottle-neck in the critical path for distributed training workloads, incurring larger and larger costs as both model and world sizes grow.
+One excellent strategy for offsetting this cost is to checkpoint in parallel, asynchronously. Below, we expand the save example
+from the `Getting Started with Distributed Checkpoint Tutorial <https://github.com/pytorch/tutorials/blob/main/recipes_source/distributed_checkpoint_recipe.rst>`__
+to show how this can be integrated quite easily with `torch.distributed.checkpoint.async_save`.
+
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:8]: Can you add a what you will learn and Prerequistes sections? You could use [cards](https://github.com/pytorch/tutorials/blob/main/beginner_source/template_tutorial.py#L9-L23) for this. 
For example:

What you will learn
* How to use parallel checkpoints asynchronously
* Effective strategies to optimize performance

Prerequisites
* PyTorch 2.4 or later
* Any other prereqs like understanding of async_save or other

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+Notes on Asynchronous Checkpointing
+------------------------------------
+Before getting started with Asynchronous Checkpointing, it's important that we discuss some differences and limitations as compared to synchronous checkpointing.
+Speciically:
+
+* Memory requirements - Asynchronous checkpointing works by first copying models into internal CPU-buffers.
+    This is helpful since it ensures model and optimizer weights are not changing while the model is still checkpointing,
+    but does raise CPU memory by a factor of checkpoint size times the number of process on the host.
+
+* Checkpoint Management - Since checkpointing is Asynchronous, it is up to the user to manage concurrently run checkpoints. In general users can
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:19]:
```suggestion
* Checkpoint Management - Since checkpointing is asynchronous, it is up to the user to manage concurrently run checkpoints. In general, users can
```

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    employ their own management strategies by handling the future object returned form `async_save`. For most users, we recommend limiting
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:20]:
```suggestion
    employ their own management strategies by handling the future object returned form ``async_save``. For most users, we recommend limiting
```

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            run_fsdp_checkpoint_save_example,
+            args=(world_size,),
+            nprocs=world_size,
+            join=True,
+        )
+
+
+Even more performance with Pinned Memory
+-----------------------------------------
+If the above optimization is still not performant enough, users may wish to take advantage of an additional optimization for GPU models which utilizes a pinned memory buffer for checkpoint staging.
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:137]:
```suggestion
If the above optimization is still not performant enough, you can take advantage of an additional optimization for GPU models which utilizes a pinned memory buffer for checkpoint staging.
```

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+Specifically, this optimization attacks the main overhead of asynchronous checkpointing, which is the in-memory copying to checkpointing buffers. By maintaing a pinned memory buffer between
+checkpoint requests users can take advantage of direct memory access to speed up this copy.
+
+Note: The main drawback of this optimization is the persistence of the buffer in between checkpointing steps. Without the pinned memory optimization (as demonstrated above),
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:141]:
```suggestion
.. note:: The main drawback of this optimization technique is the persistence of the buffer between checkpointing steps. Without the pinned memory optimization (as demonstrated above),
```

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+any checkpointing buffers are released as soon as checkpointing is finished. With the pinned memory implementation, this buffer is maintained in between steps, leading to the same
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:142]:
```suggestion
any checkpointing buffers are released as soon as checkpointing is finished. With the pinned memory implementation, this buffer is maintained between steps, leading to the same
```

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+    if __name__ == "__main__":
+        world_size = torch.cuda.device_count()
+        print(f"Running fsdp checkpoint example on {world_size} devices.")
+        mp.spawn(
+            run_fsdp_checkpoint_save_example,
+            args=(world_size,),
+            nprocs=world_size,
+            join=True,
+        )
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:259]: Can we please add a short conclusion? Something like: 

Conclusion
-----------

In this tutorial, we have learned that Asynchronous Saving with Distributed Checkpoint (DCP) offers an efficient solution to the bottleneck challenges in distributed training. By implementing asynchronous operations, it enhances computational resource utilization, reducing the time and memory costs associated with traditional synchronous checkpointing.

See also:
* Link 1
* Link 2 
* ...

File: recipes_source/distributed_checkpoint_recipe.rst [hunk 8]
```diff
@@ -161,6 +187,36 @@ The reason that we need the ``state_dict`` prior to loading is:
     CHECKPOINT_DIR = "checkpoint"
 
 
+    class AppState(Stateful):
```
Comment (svekars) [inline recipes_source/distributed_checkpoint_recipe.rst:190]: Isn't this the same as lines 70-99?

File: recipes_source/distributed_checkpoint_recipe.rst [hunk 9]
```diff
@@ -274,6 +326,37 @@ the intent is to save or load in "non-distributed" style, meaning entirely in th
         run_checkpoint_load_example()
 
 
+Formats
+----------
+One drawback not yet mentioned is that DCP saves checkpoints in a format which is inherently different then those generated using torch.save.
+Since this can be an issue when users wish to share models with users used to the torch.save format, or in general just want to add format flexibility
+to their applications. For this case, we provide the `format_utils` module in `torch.distributed.checkpoint.format_utils`.
```
Comment (svekars) [inline recipes_source/distributed_checkpoint_recipe.rst:333]:
```suggestion
to their applications. For this case, we provide the ``format_utils`` module in ``torch.distributed.checkpoint.format_utils``.
```

File: recipes_source/distributed_checkpoint_recipe.rst [hunk 10]
```diff
@@ -274,6 +326,37 @@ the intent is to save or load in "non-distributed" style, meaning entirely in th
...
+
+A command line utility is provided for the users convenience, which follows the following format:
+`python -m torch.distributed.checkpoint.format_utils -m <checkpoint location> <location to write formats to> <mode>` where mode is one of `torch_to_dcp` or `dcp_to_torch`.
```
Comment (svekars) [inline recipes_source/distributed_checkpoint_recipe.rst:336]:
```suggestion
.. code-block:: bash

   python -m torch.distributed.checkpoint.format_utils -m <checkpoint location> <location to write formats to> <mode> 
   
In the command above,  ``mode`` is either ``torch_to_dcp`` or ``dcp_to_torch``.
```

**Review by svekars — COMMENTED** [review]
Some editorial suggestions - let me know if you have any questions.

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+
+Notes on Asynchronous Checkpointing
+------------------------------------
+Before getting started with Asynchronous Checkpointing, it's important that we discuss some differences and limitations as compared to synchronous checkpointing.
+Speciically:
+
+* Memory requirements - Asynchronous checkpointing works by first copying models into internal CPU-buffers.
+    This is helpful since it ensures model and optimizer weights are not changing while the model is still checkpointing,
+    but does raise CPU memory by a factor of checkpoint size times the number of process on the host.
```
Comment (fegin) [inline recipes_source/distributed_async_checkpoint_recipe.rst:17]: I feel, this `checkpoint size times the number of process on the host` is ambiguous. 1) checkpoint size, is this the global checkpoint size or per-GPU checkpoint size? It is actually the per-GPU checkpoint size. 2) the number of process is also ambiguous. Can we just say the number of GPUs per host?

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+Even more performance with Pinned Memory
+-----------------------------------------
+If the above optimization is still not performant enough, users may wish to take advantage of an additional optimization for GPU models which utilizes a pinned memory buffer for checkpoint staging.
+Specifically, this optimization attacks the main overhead of asynchronous checkpointing, which is the in-memory copying to checkpointing buffers. By maintaing a pinned memory buffer between
+checkpoint requests users can take advantage of direct memory access to speed up this copy.
+
+Note: The main drawback of this optimization is the persistence of the buffer in between checkpointing steps. Without the pinned memory optimization (as demonstrated above),
+any checkpointing buffers are released as soon as checkpointing is finished. With the pinned memory implementation, this buffer is maintained in between steps, leading to the same
+peak memory pressure being sustained through the application life.
```
Comment (fegin) [inline recipes_source/distributed_async_checkpoint_recipe.rst:143]: This somewhat mislead the main issue of using pinned memory. The main issue of using pinned memory is that CUDA driver has to allocate ``page-lock`` memory which is scarce compared to ``pageable`` memory.

**Review by wz337 — APPROVED** [review]
Other than the editorial change Svetlana suggested, the content looks good to me. Stamping to unblock.

File: recipes_source/distributed_checkpoint_recipe.rst (hunk 8 above)
Comment (LucasLLC) [reply recipes_source/distributed_checkpoint_recipe.rst:190]: Yep @svekars , I thought it would be better if each code block was self contained. If that's not the case, I can remove duplicates

Comment (svekars) [reply recipes_source/distributed_checkpoint_recipe.rst:190]: I'll leave it up to you to decide. I just wanted to make sure it wasn't copy pasted twice by mistake.

**Comment by LucasLLC** [comment]
Updated all comments!

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 13]
```diff
@@ -0,0 +1,287 @@
+Asynchronous Saving with Distributed Checkpoint (DCP)
+=====================================================
+
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:3]:
```suggestion

**Author:** `Lucas Pasqualin <https://github.com/lucasllc>`__, `Iris Zhang <https://github.com/wz337>`__, `Rodrigo Kumpera <https://github.com/kumpera>`__, `Chien-Chin Huang <https://github.com/fegin>`__

```

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+Checkpointing is often a bottle-neck in the critical path for distributed training workloads, incurring larger and larger costs as both model and world sizes grow.
+One excellent strategy for offsetting this cost is to checkpoint in parallel, asynchronously. Below, we expand the save example
+from the `Getting Started with Distributed Checkpoint Tutorial <https://github.com/pytorch/tutorials/blob/main/recipes_source/distributed_checkpoint_recipe.rst>`__
+to show how this can be integrated quite easily with ``torch.distributed.checkpoint.async_save``.
+
+**Author**: , `Lucas Pasqualin <https://github.com/lucasllc>`__, `Iris Zhang <https://github.com/wz337>`__, `Rodrigo Kumpera <https://github.com/kumpera>`__, `Chien-Chin Huang <https://github.com/fegin>`__
```
Comment (svekars) [inline recipes_source/distributed_async_checkpoint_recipe.rst:9]: Moved to the line above
```suggestion
```

**Review by svekars — COMMENTED** [review]
Formatting cleanup. Otherwise, LGTM!

**Comment by LucasLLC** [comment]
> Formatting cleanup. Otherwise, LGTM! Please do not merge until 1 week before the release (July 17).

Thank you so much for all the editing help! We should add you as an author :)

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+       :class-card: card-prerequisites
+
+       * PyTorch v2.4.0 or later
+       * `Getting Started with Distributed Checkpoint Tutorial <https://github.com/pytorch/tutorials/blob/main/recipes_source/distributed_checkpoint_recipe.rst>`__
+
+
+Asynchronous Checkpointing Overview
+------------------------------------
+Before getting started with Asynchronous Checkpointing, it's important to understand it's differences and limitations as compared to synchronous checkpointing.
+Speciically:
```
Comment (lessw2020) [inline recipes_source/distributed_async_checkpoint_recipe.rst:30]: spelling error = "Specifically:"

**Review by svekars — APPROVED** [review]
